From 24947051435f45ae80cd650b848a8914feb25998 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Hannes=20M=C3=BCller?= <>
Date: Mon, 17 Oct 2022 21:22:19 +0200
Subject: [PATCH] Try windows conversion of %load-path on MinGW

Try using cygpath to convert SCM_LIBRARY_DIR from posix to windows path name
for initial %load-path. This shall allow to run guile without explicitly
setting GUILE LOAD_PATH.
diff --git a/libguile/load.c b/libguile/load.c
index 5ca4e07..5860b8a 100644
--- a/libguile/load.c
+++ b/libguile/load.c
@@ -206,9 +206,33 @@ scm_init_load_path ()
   SCM path = SCM_EOL;

 #ifdef SCM_LIBRARY_DIR
+# ifdef __MINGW32__
+  /* try conversion of posix SCM_LIBRARY_DIR to windows path using cygpath */
+  char *char_ptr, output[260];
+  int exit_val;
+  FILE *fp;
+  SCM cmd;
+
+  cmd = scm_string_append (scm_list_3 (
+                             scm_from_locale_string ("cygpath -w "),
+                             scm_from_locale_string (SCM_LIBRARY_DIR),
+			     scm_from_locale_string (" 2>&1")));
+
+  fp = popen (scm_to_locale_string(cmd), "r");
+  if (fp != NULL) {
+    char_ptr = fgets (output, sizeof(output), fp);
+    exit_val = pclose (fp);
+    if(!exit_val && char_ptr != NULL && strlen (output) > 0 &&
+       strlen (output) != sizeof (output)-1) {
+      output[strlen (output)-1] = '\0'; /* remove newline */
+      path = scm_list_1 (scm_from_locale_string (output));
+    }
+  }
+# else
   path = scm_list_3 (scm_from_locale_string (SCM_SITE_DIR),
		     scm_from_locale_string (SCM_LIBRARY_DIR),
		     scm_from_locale_string (SCM_PKGDATA_DIR));
+# endif
 #endif /* SCM_LIBRARY_DIR */

   env = getenv ("GUILE_LOAD_PATH");
